datafusion/catalog/src/catalog.rs

```diff
@@ -102,24 +103,25 @@ use datafusion_common::Result;
 ///
 /// [`TableProvider`]: crate::TableProvider
 
+#[async_trait]
 pub trait CatalogProvider: Debug + Sync + Send {
     /// Returns the catalog provider as [`Any`]
     /// so that it can be downcast to a specific implementation.
     fn as_any(&self) -> &dyn Any;
 
     /// Retrieves the list of available schema names in this catalog.
-    fn schema_names(&self) -> Vec<String>;
+    async fn schema_names(&self) -> Vec<String>;
```

Comment:
Probably the latter, not sure about whether it should be `'_` or `'static`

Reply:
I'll try and get away with `'_`.  Otherwise the memory provider will need to clone its contents.

If callers need static they can Arc the provider, clone it, and wrap with the call.

Reply:
The reason to perhaps prefer `'static` is python compatibility - https://github.com/apache/arrow-rs/issues/6587

Reply:
Yeah, I ran into a similar problem as well.  It's bothersome I can't do...

```
let catalog_list = self.catalog_list.clone()
let catalog_names = catalog_list.catalog_names().await;
MagicBindingStream::new(catalog_list, catalog_names)
```

Similar to what I can do with futures...

```
let temporary_obj = self.thing.clone();
async move {
  temporary_obj.fut_returning_that_depends_on_self().await
}
```

Reply:
Ok.  How about we retreat to:

```
async fn schema_names(&self) -> Result<Box<dyn Iterator<Item = String>>>;
```

with the rationale that real world implementations should be caching results anyways.  Note there is once again an implicit `'_` bound on the result but since we're returning a future and not a stream it's easily removed if needed.

Reply:
If there is a case where there are so many tables / schemas / catalogs that caching the full list is too expensive then the planner is probably going to have other problems anyways.  I'm guessing the planner is iterating through this list and constructing a hash map it uses to resolve lookups and so the thing is going to get cached one way or another.

Reply:
That seems like the worst of both worlds 😅

What is the issue with a static stream, its fairly extensively used in DF?

Reply:
> What is the issue with a static stream, its fairly extensively used in DF?

Just me being foolish.  Having now gone full circle on the iterator approach I'm ready to do static stream again :face_exhaling:

Reply:
Ok.  I have switched to `BoxStream<'static, Result<String>>`